src/lib/optimizer/chunk_statistics.hpp

```diff
@@ -0,0 +1,165 @@
+    switch (predicate_type) {
+      case PredicateCondition::GreaterThan:
+        return t_value >= _max;
+      case PredicateCondition::GreaterThanEquals:
+        return t_value > _max;
+      case PredicateCondition::LessThan:
+        return t_value <= _min;
+      case PredicateCondition::LessThanEquals:
+        return t_value < _min;
+      case PredicateCondition::Equals:
+        return t_value < _min || t_value > _max;
+      default:
+        return false;
+    }
+  }
+ protected:
+  T _min;
+  T _max;
+};
+
+template <typename T>
+class RangeFilter : public BaseFilter {
+ public:
+  RangeFilter(std::vector<std::pair<T,T>> ranges) : _ranges(ranges) {};
+  virtual ~RangeFilter() = default;
+
+  static std::shared_ptr<RangeFilter<T>> build_filter(const pmr_vector<T>& dictionary);
+  
+  bool can_prune(const AllTypeVariant& value, const PredicateCondition predicate_type) const override {
+    T t_value = type_cast<T>(value);
+    // Operators work as follows: value_from_table <operator> t_value
+    // e.g. OpGreaterThan: value_from_table > t_value
+    // thus we can exclude chunk if t_value >= _max since then no value from the table can be greater than t_value
+    switch (predicate_type) {
+      case PredicateCondition::Equals: {
+        bool prunable = false;
+        for(const auto& bounds : _ranges) {
+          auto& [min, max] = bounds;
+          prunable |= t_value > min && t_value < max;
+        }
+        return prunable;
+      }
+      default:
+        return false;
+    }
+  }
+
+ protected:
+  std::vector<std::pair<T,T>> _ranges; 
+};
+
+template<typename T>
+std::shared_ptr<RangeFilter<T>> RangeFilter<T>::build_filter(const pmr_vector<T>& dictionary) {
+  if constexpr (std::is_same<T, std::string>::value) {
+      return nullptr;
+  } else {
+      // calculate distances by taking the difference between two neighbouring elements
+      std::vector<std::pair<T, size_t>> distances;
+      for (auto dict_it = dictionary.cbegin(); dict_it + 1 != dictionary.cend(); ++dict_it) {
+        auto dict_it_next = dict_it + 1;
+        distances.emplace_back(std::make_pair(*dict_it_next - *dict_it, std::distance(dictionary.cbegin(), dict_it)));
+      }
+
+      std::sort(distances.begin(), distances.end(),
+                [](const auto& pair1, const auto& pair2){ return pair1.first > pair2.first; });
+
+      // select how many ranges we want in the filter
+      // make this customizable?
+      const size_t max_ranges_count = 10;
+
+      if(max_ranges_count - 1 < distances.size()) {
+        distances.erase(distances.cbegin() + (max_ranges_count - 1), distances.cend());
+      }
+
+      std::sort(distances.begin(), distances.end(),
+                [](const auto& pair1, const auto& pair2){ return pair1.second < pair2.second; });
+
+      // derive intervals where items don't exist from distances
+      //
+      //         index  index + 1
+      //         v      v
+      // 1 2 3 4 5      10 11     15 16
```

Comment:
@Bouncner - Could you review the algorithmic part of this?

Reply:
I am actually not too happy with using the dictionary here. The purpose of using the chunkEncode methods by Max was that we do not make assumptions about the actual column compression.